src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,217 @@
+ *
+ * <p>
+ *     Result list
+ * </p>
+ * <ul>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS/LITERAL_PUBLIC
+ *     </li>
+ * </ul>
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathQueryGenerator {
+    /** The root ast. */
+    private final DetailAST rootAst;
+    /** The line number. */
+    private final int lineNumber;
+    /** The column number. */
+    private final int columnNumber;
+
+    /**
+     * Creates a new {@code XpathQueryGenerator} instance.
+     *
+     * @param rootAst root ast
+     * @param lineNumber line number
+     * @param columnNumber column number
+     */
+    public XpathQueryGenerator(DetailAST rootAst, int lineNumber, int columnNumber) {
+        this.rootAst = rootAst;
+        this.lineNumber = lineNumber;
+        this.columnNumber = columnNumber;
+    }
+
+    /**
+     * Returns list of xpath queries of nodes, matching line and column number.
+     * This approach uses traversion through DetailAst tree.
+     * @return list of xpath queries of nodes, matching line and column number
+     */
+    public List<String> generate() {
+        final List<DetailAST> matchingAstElements = getMatchingAstElements(rootAst);
+        final List<String> generated = new ArrayList<>();
+        for (DetailAST ast : matchingAstElements) {
+            String xpathQuery = getXpathQuery(null, ast);
+            if (!isUniqueAst(ast)) {
+                final DetailAST child = findChildWithIdent(ast);
+                if (child != null) {
+                    xpathQuery = xpathQuery + "[." + getXpathQuery(ast, child) + "]";
+                }
+            }
+            generated.add(xpathQuery);
+        }
+        return generated;
+    }
+
+    /**
+     * Returns child {@code DetailAst} element of the given root,
+     * which has child element with token type equals to {@link TokenTypes#IDENT}.
+     * @param root {@code DetailAST} root ast
+     * @return child {@code DetailAst} element of the given root
+     */
+    private DetailAST findChildWithIdent(DetailAST root) {
+        DetailAST curNode = root;
+        while (curNode.findFirstToken(TokenTypes.IDENT) == null) {
+            DetailAST toVisit = curNode.getFirstChild();
+            while (toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            if (curNode == root.getParent()) {
+                curNode = null;
+                break;
+            }
+
+            curNode = toVisit;
+        }
+        return curNode;
+    }
+
+    /**
+     * Finds if {@code DetailAst} element is unique, across all elements or across siblings.
```

Comment:
Describe all algorithm of the method in the documentation. I do not really understand the purpose of the method without debugging it